source/TextReplacements.cpp

```diff
@@ -0,0 +1,73 @@
+/* TextReplacements.cpp
+Copyright (c) 2021 by Amazinite
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#include "TextReplacements.h"
+
+#include "ConditionSet.h"
+#include "DataNode.h"
+#include "PlayerInfo.h"
+
+#include <set>
+
+using namespace std;
+
+
+
+void TextReplacements::Load(const DataNode &node)
+{
+	// Check for reserved keys. Only some hardcoded replacement keys are
+	// reserved, as these ones are done on the fly after all other replacements
+	// have been done.
+	set<string> reserved = {"<first>", "<last>", "<ship>"};
+	
+	for(const DataNode &child : node)
+	{
+		if(child.Size() < 2)
+			child.PrintTrace("Skipping improper substitution syntax:");
+		else
+		{
+			const string &key = child.Token(0);
```

Comment:
I think it makes sense to emit a warning if the key doesn't have the `<>` required (and also add them if they're missing).

Reply:
I don't think we should enforce all substitutions to be of that form. That's what the game uses for hardcoded replacements and that's what I imagine will be used a majority of the time, but I don't see a reason why it would be necessary to enforce all substitutions to be that way. One good example I can think of is if a content creator wants to use the name of an already existing substitution without overriding it. For example, perhaps they have substitutions with keys `{first}`, `{second}`, `{third}` to avoid clashing with the already existing `<first>` replacement.

Reply:
I understand the humorous aspect...
```
substitutions
	"captain" "fool"
	"Captain" "Fool"
	"Hai" "Squirrel"

substitutions
	" " "-"
	" " "."
		day % 2 == 0
	" " ""
		day % 4 == 0
```
But I would prefer the current PR to only accept the version where `<>` is required. We can then still do a follow-up PR where we can replace any text we desire, but being able to replace any possible word feels a bit out of scope for now.

Reply:
> We can then still do a follow-up PR where we can replace any text we desire, but being able to replace any possible word feels a bit out of scope for now.

I'm not sure why it'd make sense to allow the replacing of any text later, since it seems to me like allowing that would just be a matter of removing the very code needed to enforce the style of using `<key>` for replacements.

Reply:
Because I expect a number of additional requirements on that follow-up PR besides just reverting the check on '<>' markers. For example safeguards against partial-word matching (whole-word-matching) if we don't have the markers.

Suppose somebody wants to replace all occurences of `depart` with `launch`. Then we don't want that text replacement to happen in a text like `"Passengers Jan and Piet, Please report to departure desk 5."`.

Reply:
Let's first try to get the simple feature of user-defined `<text replacements>` merged before we discuss what is exactly needed to get an `any text replacement` feature merged.

Reply:
> Suppose somebody wants to replace all occurences of `depart` with `launch`. Then we don't want that text replacement to happen in a text like `"Passengers Jan and Piet, Please report to departure desk 5."`.

Well that could be done manually already under the current implementation.
```
substitutions
	"depart " "launch "
	"depart," "launch,"
	"depart;" "launch;"
	"depart:" "launch:"
	"depart." "launch."
	"depart!" "launch!"
	"depart?" "launch?"
```
I think that automatic whole-word-matching would definitely be a good followup PR, but I don't think that means we should disallow manually doing that now. (Although a syntax for whole-word-matching might be a good idea to brainstorm now, and we could potentially reserve the syntax until it's implemented. For example, perhaps a key `=depart=` indicates whole-word-matching, and we disallow the use of any key of that format in this PR.)

Reply:
Not sure why it hasn't been mentioned yet but currently the substitution algorithm (`Format::Replace`) *requires* the use of `<>` markers. So you definitely need to change it if you want to support substitutions without them. It will also make the substitution algorithm more expensive to run.

Reply:
We could, but to be complete one would then also need to check for the beginning of the word (and add Capitalized versions of the word). Otherwise we would still get support questions when somebody comes up with texts like this:
```
	"A backsidepart is missing."
	"What is the zipcodepart of the address?"
	"Did you see the codepart that was written by the programmer?"
```
If we are going to allow all replacements, then I would suggest to at least add a warning when we use non-marked replacements (as quyykk also suggested).

Reply:
> Not sure why it hasn't been mentioned yet but currently the substitution algorithm (`Format::Replace`) *requires* the use of `<>` markers.

I know that Format::Replace handled the actual replacements, but that's not something I was aware of. Thanks.

Reply:
Does the latest commit look good?